### ISSUE TITLE
Kolibri 0.15.x running on Windows 10 - Python high CPU utilization

### ISSUE THREAD

**Issue by pcenov** [learningequality/kolibri#9908]
## Observed behavior
In Windows 10 one can notice that when running Kolibri 0.15 this is causing high CPU utilization of up to 35-50% even when Kolibri is not in use.
This was initially internally discussed here: https://learningequality.slack.com/archives/CAUTAU14Y/p1665777085135879
And is now reported in the community forum here: https://community.learningequality.org/t/python-high-cpu-utilization/2665

## Expected
It should be further investigating what is the root cause of the issue and the CPU usage should be minimized.

## Steps to reproduce the issue
1. Install Kolibri v0.15.x on a Windows 10 device.
2. Monitor the Python CPU usage.

## Screenshots
![1c1de90d27ae9bb73b34521f76c85d8a470481cf](https://user-images.githubusercontent.com/79847249/207094052-43344157-c3e2-4f6d-88bb-50be8223683b.png)

## Usage Details
Windows 10 and Ubuntu 20 - Chrome, Firefox

**Comment by pcenov** [comment]
@radinamatic

**Comment by radinamatic** [comment]
Suggested by @rtibbles:

> The possibilities that come to mind are:
> * The background task runner is somehow running amock.
> * The magicbus thread manager is running amock.
> * Something funny is happening with the pingback task.
> * Version of Python used if installer is signed by Github action vs manually in the signer VM, and whether that might mean that the Python cryptography package had its C extensions installed or not.

cc @marcellamaki for planned patches and iteration planning

**Comment by marcellamaki** [comment]
@pcenov and @radinamatic - as you begin integration testing for 0.16, can you confirm if this continues to be a problem on windows 10 for the newer release? Thank you!

**Comment by pcenov** [comment]
Hi @marcellamaki - I confirm that this is still the same in [Beta 7](https://github.com/learningequality/kolibri/releases/tag/v0.16.0-beta7):

![2023-11-02_18-30-59](https://github.com/learningequality/kolibri/assets/79847249/b03bd2bb-5b3f-445f-995a-e68e52091105)

**Comment by Dimi20cen** [comment]
Hi,

This issue is still present in Kolibri `0.17.5` on Windows. I was experiencing a consistent ~25% CPU usage from the Python process, even when Kolibri was completely idle.

It's a busy-wait loop in the `magicbus` library, which Kolibri uses for its process management.

The issue is in the `_wait()` function within the `Bus.wait` method in `kolibri/dist/magicbus/base.py`. On Windows, the `select.select()` call is not blocking as expected, causing the `while` loop to spin and consume a lot of the cpu.

I used `py-spy` to profile the running Kolibri process, which showed that ~96% of the CPU time was being spent in this `_wait` function.

<img width="737" height="456" alt="Image" src="https://github.com/user-attachments/assets/c66823f8-9e4e-4ecc-9cd2-671b14cfe12d" />


The issue was resolved by introducting `time.sleep(0.01)` call into the loop.
In `kolibri/dist/magicbus/base.py`:
```
        def _wait():
            try:
                while self.state not in _states_to_wait_for:
                    time.sleep(0.01) <----- HERE
                    if select:
                        try:
                            r, w, x = select.select([read_fd], [], [], interval)
                            if r:
```

**Comment by radinamatic** [comment]
Is this something we can change, fix and merge independently of the new Windows installer coming from the GSoC project (assuming it's not applicable and/or already fixed there 🙂)?

**Comment by Dimi20cen** [comment]
Yes, this is completely independent of the kolibri app. 

But currently, I honestly have no idea on how to fix this issue on the Kolibri repo.

The file that needs to be changed, `kolibri/dist/magicbus/base.py` is not present in the Kolibri repository. It's a build artifact, and its created during the installation process. Therefore it's not something we can just change (except if we forked the magibus repo, i guess)

Sorry if this answer is a bit confusing (I'm a bit confused as well)

**Comment by rtibbles** [comment]
Yeah, so this is definitely a bug in the underlying magicbus library. It seems to have been written when the `select` module was only available on POSIX systems, and not on Windows - however, I am also completely unable to find a version of Python where this was actually true, so this just seems like a long standing bug in magicbus.

**Comment by rtibbles** [comment]
I have filed an issue upstream, but I will fix this locally so that we are not blocked.

https://github.com/cherrypy/magicbus/issues/25

**Comment by rtibbles** [comment]
OK, I am able to replicate this issue on Linux now by faking the Windows select behaviour.

**Comment by rtibbles** [comment]
Fixed in #13673
